libs/ktxreader/include/ktxreader/Ktx2Reader.h

```diff
@@ -83,19 +91,111 @@ class Ktx2Reader {
          *      in the KTX2 blob. If they do not match, this method fails.
          *   2) It is used as a filter when determining the final internal format.
          */
-        Texture* load(const uint8_t* data, size_t size, TransferFunction transfer);
+        Texture* load(const void* data, size_t size, TransferFunction transfer);
+
+        /**
+         * Asynchronous Interface
+         * ======================
+         *
+         * Alternative API suitable for asynchronous transcoding of mipmap levels.
+         * If unsure that you need to use this, then don't, just call load() instead.
+         * Usage pseudocode:
+         *
+         *    auto async = reader->asyncCreate(data, size, TransferFunction::LINEAR);
+         *    mTexture = async->getTexture();
+         *    auto backgroundThread = spawnThread({ async->doTranscoding(); })
+         *    backgroundThread.wait();
+         *    async->uploadImages();
+         *    reader->asyncDestroy(async);
+         *
+         * In the documentation comments, "foreground thread" refers to the thread that the
+         * Filament Engine was created on.
+         */
+        class Async {
+        public:
+            /**
+             * Retrieves the Texture object.
+             *
+             * The texture is available immediately, but does not have its miplevels ready until
+             * after doTranscoding() and the subsequent uploadImages() have been completed. The
+             * caller has ownership over this texture and is responsible for freeing it after all
+             * miplevels have been uploaded.
+             */
+            Texture* getTexture() const noexcept;
+
+            /**
+             * Loads all mipmaps from the KTX2 file and transcodes them to the resolved format.
+             *
+             * This does not return until all mipmaps have been transcoded. This is typically
+             * called from a background thread.
+             */
+            Result doTranscoding();
+
+            /**
+             * Uploads pending mipmaps to the texture.
+             *
+             * This can safely be called while doTranscoding() is still working in another thread.
+             * Since this calls Texture::setImage(), it should be called from the foreground thread;
+             * see "Thread safety" in the documentation for filament::Engine.
+             */
+            void uploadImages();
```

Comment:
now I'm wondering if `uploadImages` should return whether there is more?

Reply:
The caller already knows, because they know if `doTranscoding()` has completed or not.  Note that the Ktx2Reader implementation does not create threads, it's the caller's responsibility to do so if they wish.  I think this is a reasonable separation of concerns.  (The caller in this case will be the gltfio TextureProvider that I'm adding in my next PR.)